bson/src/main/org/bson/Vector.java

```diff
@@ -0,0 +1,268 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.bson;
+
+
+import java.util.Arrays;
+import java.util.Objects;
+
+import static org.bson.assertions.Assertions.assertNotNull;
+import static org.bson.assertions.Assertions.isTrue;
+import static org.bson.assertions.Assertions.isTrueArgument;
+import static org.bson.assertions.Assertions.notNull;
+
+/**
+ * Represents a vector that is stored and retrieved using the BSON Binary Subtype 9 format.
+ * This class supports multiple vector {@link Dtype}'s and provides static methods to create
+ * vectors and methods to retrieve their underlying data.
+ * <p>
+ * Vectors are densely packed arrays of numbers, all the same type, which are stored efficiently
+ * in BSON using a binary format.
+ *
+ * @mongodb.server.release 6.0
+ * @see BsonBinary
+ * @since BINARY_VECTOR
+ */
+public final class Vector {
```

Comment:
Instead of having unused fields depending on the type of the vector, should we instead 
have private subclasses of Vector, one for each type, that has only the fields needed for that type?

Another option would be to make the subclasses public, and put the "as" conversion methods on the subclasses.  This would be more similar to the design of the `BsonValue` type hierarchy.

Reply:
After reconsidering, creating a subclass for each type leads to a more user-friendly API. Initially, I was concerned about class proliferation, given the number of potential vector types in the future:

```
INT4, INT8, INT16, INT32, INT64
UINT8, UINT16, UINT32, UINT64
FLOAT16, FLOAT32, FLOAT64
BOOLEAN
PACKED_BIT
DECIMAL128, DECIMAL32, DECIMAL64
COMPLEX64, COMPLEX128
```

However, having a subclass per type helps keep the logic isolated where it belongs. For example, padding isn’t applicable to float32/64 vectors or Int8/16/32/64 vectors, so having particular subclass without padding makes the API more user-friendly and focused. This is especially useful when using `PojoCodec`, where users can directly specify `Float32Vector` or `PackedBitVector`, improving clarity and usability, similar to how we handle public subclasses in `BsonValue`.

I've updated the implementation to use subclasses. Please, let me know your thoughts.

Reply:
I like it.